packages/no-trapping-shim/src/no-trapping-pony.js

```diff
@@ -0,0 +1,206 @@
+ * @param {boolean} shouldThrow
+ * @returns {boolean}
+ */
+const suppressTrappingInternal = (specimen, shouldThrow) => {
+  if (noTrappingSet.has(specimen)) {
+    return true;
+  }
+  freeze(specimen);
+  if (!proxyHandlerMap.has(specimen)) {
+    noTrappingSet.add(specimen);
+    return true;
+  }
+  const [target, handler] = proxyHandlerMap.get(specimen);
+  if (isNoTrappingInternal(target, shouldThrow)) {
+    noTrappingSet.add(specimen);
+    return true;
+  }
+  const trap = handler.suppressTrapping;
+  if (trap === undefined) {
+    const result = suppressTrappingInternal(target, shouldThrow);
+    if (result) {
+      noTrappingSet.add(specimen);
+    }
+    return result;
+  }
+  const result = apply(trap, handler, [target]);
+  const ofTarget = isNoTrappingInternal(target, shouldThrow);
+  if (result !== ofTarget) {
+    if (shouldThrow) {
+      throw TypeError(
+        `'suppressTrapping' proxy trap does not reflect 'isNoTrapping' of proxy target (which is '${ofTarget}')`,
+      );
+    }
+    return false;
+  }
+  if (result) {
+    noTrappingSet.add(specimen);
+  }
+  return result;
+};
+
+export const extraReflectMethods = freeze({
+  isNoTrapping(target) {
+    if (isPrimitive(target)) {
+      throw TypeError('Reflect.isNoTrapping called on non-object');
+    }
+    return isNoTrappingInternal(target, false);
+  },
+  suppressTrapping(target) {
+    if (isPrimitive(target)) {
+      throw TypeError('Reflect.suppressTrapping called on non-object');
+    }
+    return suppressTrappingInternal(target, false);
+  },
+});
+
+export const extraObjectMethods = freeze({
+  isNoTrapping(target) {
+    if (isPrimitive(target)) {
+      return true;
+    }
+    return isNoTrappingInternal(target, true);
+  },
+  suppressTrapping(target) {
+    if (isPrimitive(target)) {
+      return target;
+    }
+    if (suppressTrappingInternal(target, true)) {
+      return target;
+    }
+    throw TypeError('preventExtensions trap returned falsy');
+  },
+});
+
+const addExtras = (base, ...extrasArgs) => {
+  for (const extras of extrasArgs) {
+    for (const key of ownKeys(extras)) {
+      if (base[key] !== extras[key]) {
+        defineProperty(base, key, {
+          value: extras[key],
+          writable: true,
+          enumerable: false,
+          configurable: true,
+        });
+      }
+    }
+  }
+};
+
+const ReflectPlus = {};
+addExtras(ReflectPlus, OriginalReflect, extraReflectMethods);
+export { ReflectPlus };
+
+const ObjectPlus = function Object(...args) {
+  if (new.target) {
+    return construct(OriginalObject, args, new.target);
+  } else {
+    return apply(OriginalObject, this, args);
+  }
+};
+ObjectPlus.prototype = OriginalObject.prototype;
+addExtras(ObjectPlus, OriginalObject, extraObjectMethods);
+export { ObjectPlus };
+
+const makeMetaHandler = handler =>
+  freeze({
+    get(_, trapName, _receiver) {
+      return freeze((target, ...rest) => {
+        if (
+          isNoTrappingInternal(target, true) ||
```

Comment:
Isn't setting the proxy as non-trapping effectively the same as removing all traps from the handler? Since we interject on the user provided handler, can't we actually do just that?

Reply:
Can't do it immediately, since a target becoming non-trapping does not immediately reveal all the proxies with that target that then need to become non-trapping.

Rather, in making the metaHandler lazily override the trapName on the handlerPlus, I also arranged for it to lazily re-override with `undefined` for the no-trapping case. So after the first time, handlerPlus[trapName] will be `undefined` and it will continue to do the right thing, but faster.

Done.